libfoundation/src/foundation-cf.cpp

```diff
@@ -0,0 +1,470 @@
+    if (t_cf_data == nullptr)
+    {
+        return false;
+    }
+    r_cf_data = t_cf_data;
+    return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+MC_DLLEXPORT_DEF
+bool MCProperListCreateWithCFArrayRef(CFArrayRef p_cf_array, bool p_use_lists, MCProperListRef& r_list)
+{
+    MCAutoProperListRef t_list;
+    if (!MCProperListCreateMutable(&t_list))
+    {
+        return false;
+    }
+    
+    for(CFIndex t_index = 0; t_index < CFArrayGetCount(p_cf_array); t_index++)
+    {
+        MCAutoValueRef t_value;
+        if (!MCValueCreateWithCFTypeRef(CFArrayGetValueAtIndex(p_cf_array, t_index), p_use_lists, &t_value) ||
+            !MCProperListPushElementOntoBack(*t_list, *t_value))
+        {
+            return false;
+        }
+    }
+    
+    if (!t_list.MakeImmutable())
+    {
+        return false;
+    }
+    
+    r_list = t_list.Take();
+    
+    return true;
+}
+
+MC_DLLEXPORT_DEF
+bool MCProperListConvertToCFArrayRef(MCProperListRef p_list, bool p_use_lists, CFArrayRef& r_cf_array)
+{
+    MCAutoCustomPointer<const void, CFDeleter> t_cf_array =
+            CFArrayCreateMutable(nullptr, MCProperListGetLength(p_list), &kCFTypeArrayCallBacks);
+    
+    if (!t_cf_array)
+    {
+        return MCErrorThrowOutOfMemory();
+    }
+    
+    for(uindex_t t_index = 0; t_index < MCProperListGetLength(p_list); t_index++)
+    {
+        MCAutoCustomPointer<const void, CFDeleter> t_cf_value;
+        if (!MCValueConvertToCFTypeRef(MCProperListFetchElementAtIndex(p_list, t_index), p_use_lists, &t_cf_value))
+        {
+            return false;
+        }
+        CFArraySetValueAtIndex((CFMutableArrayRef)*t_cf_value, t_index, *t_cf_value);
+    }
+    
+    r_cf_array = (CFArrayRef)t_cf_array.Release();
+    
+    return true;
+}
+
+MC_DLLEXPORT_DEF
+bool MCProperListCreateWithCFDictionaryRef(CFDictionaryRef p_cf_dictionary, bool p_use_lists, MCProperListRef& r_proper_list)
+{
+    MCAutoArrayRef t_array;
+    if (!MCArrayCreateWithCFDictionaryRef(p_cf_dictionary, p_use_lists, &t_array))
+    {
+        return false;
+    }
+    
+    return MCArrayConvertToProperList(*t_array, r_proper_list);
+}
+
+MC_DLLEXPORT_DEF
+bool MCProperListConvertToCFDictionaryRef(MCProperListRef p_list, bool p_use_lists, CFDictionaryRef& r_cf_dictionary)
+{
+    MCAutoArrayRef t_array;
+    if (!MCProperListConvertToArray(p_list, &t_array))
+    {
+        return false;
+    }
+    
+    return MCArrayConvertToCFDictionaryRef(*t_array, p_use_lists, r_cf_dictionary);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+MC_DLLEXPORT_DEF
+bool MCArrayCreateWithCFDictionaryRef(CFDictionaryRef p_cf_dict, bool p_use_lists, MCArrayRef& r_array)
+{
+    MCAutoArrayRef t_array;
+    if (!MCArrayCreateMutable(&t_array))
+    {
+        return false;
+    }
+    
+    CFDictionaryApplyFunction(p_cf_dict, nullptr, nullptr);
+    
+    return false;
```

Comment:
This should probably throw some kind of unimplemented error, if it's meant to be unimplemented. Or at least a comment :-)

Reply:
Heh - or actually implement it :)